File: lab5/Dictionary.c
// 
//		Dictionary.c
// 		James Vrionis
// 		A Dictionary ADT based on the linked list data structure. Elements of the 
//      Dictionary will be pairs of Strings called key and value. (key, value): 
//      keys are distinct while values may be repeated. Implement the seven 
//      given ADT operations given from pa3 into c.
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

/* NodeObj */
typedef struct NodeObj {
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

/* NODE */
typedef NodeObj* Node;

/* newNode(): 
 * Constructor for Node type 
 */
Node newNode(char* k, char* v) {
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return(N);
} 

/* freeNode(): 
 * Destructor for Node type 
 */
void freeNode(Node* pN) {
	if(pN != NULL && *pN != NULL) {
		free(*pN);
		*pN = NULL;
	}
}

/* DictionaryObj
 * 
 */
typedef struct DictionaryObj {
	Node head;
	Node tail;
	int numItems;
} DictionaryObj;

/* newDictionary(): 
 * Dictionary Type Constructor  
 */
Dictionary newDictionary(void) {
 	Dictionary D = malloc(sizeof(DictionaryObj));
 	assert(D != NULL);
 	D->head = NULL;
 	D->tail = NULL;
 	D->numItems = 0;
 	return(D);	
}
 
 /* freeDictionary()
  * Dictionary Type Destructor
  */
void freeDictionary(Dictionary* pD) {
	if(pD != NULL && *pD != NULL) {
		if( !isEmpty(*pD))makeEmpty(*pD);
			free(*pD);
		*pD = NULL;	
	} 		
}

 /* isEmpty()
  * pre: none
  * post: return (true) 1 if S is empty, if not (false) 0 
  */
int isEmpty(Dictionary D) {
 	if(D == NULL) {
 		fprintf(stderr, "isEmpty Error: calling isEmpty on NULL Dictionary reference\n ");
 		exit(EXIT_FAILURE);
 	}
 	if(D->numItems > 0) {
 		return 0; // False
 	}
 	return 1; // True	
 }
 
 /* size()
  * pre: none
  * post: returns number of (key,value) pairs in D 
  */
int size(Dictionary D) {
	return D->numItems; 
 }

 /* lookup()
  * pre: none
  * post: Returns value v such that (k,v) in in D, or NULL if it's not there
  */
char* lookup(Dictionary D, char* k) {
	Node N = D->head;
	if(D == NULL) {
		fprintf(stderr, "Lookup Error: calling lookup on NULL Dictionary\n");
		exit(EXIT_FAILURE);
	}
	while(N != NULL) {
		if(strcmp(N->key,k)==0) return(N->value);
		N = N->next;
	}
		return(NULL);
}

 /* insert()
  * pre: none
  * post: Returns value v such that (k,v) in in D, or NULL if it's not there
  */
void insert(Dictionary D, char* k, char* v) {
	if(D-> numItems == 0) {
		D->head = D->tail = newNode(k,v);
	}else {
		Node N;
		N = newNode(k,v);
		D->tail->next = N;
		D->tail = N;
	}
	D->numItems++;
}	

 /* makeEmpty()
  * pre: none
  * post: Resets D to the Empty State.
  */
void makeEmpty(Dictionary D) {
	D->head = NULL;
	D->tail = NULL;
	D->numItems= 0; 
}
  
 /* delete()
  * pre: lookup(D,k) != NULL 
  * post: Deletes a (key,value) pair 
  */
void delete(Dictionary D, char* k) {
// Make an indirect pointer that points to *Address* of the thing well update
//	indirect = &head;
// walk the list, looking for the thing that thing that points to the entry we want to remove
// while ( (*indirect) != &key)
//    indirect = &(*indirect)->next;
// ... and just remove it.
// *indirect = entry -> next; 

	Node N = D->head;
	if(lookup(D,k) == NULL) {
		fprintf(stderr, "Delete Error: Key not found!\n");		
		exit(EXIT_FAILURE);
	}	
	if(strcmp(N->key,k)==0) {
		Node P = D->head;
		Node G = P->next;
		D->head = G;
		freeNode(&P);
	}
	else {
		while(N != NULL && N->next != NULL) {
			if(strcmp(N->next->key, k)== 0) {
				Node G = N;
				Node C = N->next;
				G->next = C->next;
				N=G;
				freeNode(&C);
			}	
		N = N->next;
		}
	}
	D->numItems--;
}

 /* printDictionary()
  * pre: none
  * post: prints text of D to file pointed to by out
  */
void printDictionary(FILE* out, Dictionary D) {
	Node N;
	if(D == NULL) {
		fprintf(stderr, "PrintDictionary Error: calling print function on NULL reference\n");
		exit(EXIT_FAILURE);	
	} 	
	for(N=D->head; N !=NULL; N=N->next) {
		fprintf(out, " %s %s\n", N->key, N->value);	
	}
 	fprintf(out,"\n");
 }

File: pa3/Dictionary.java
/* Program: Dictionary.java
 * Author:  James Michael Vrionis
 * ID    :  JVrionis
 * Date  :  11-3-16
 * Usuage:  A Dictionary ADT based on the linked list data structure. Elements of the 
 *          Dictionary will be pairs of Strings called key and value. (key, value): 
 *          keys are distinct while values may be repeated. Implement the seven 
 *          given ADT operations given.
 *
 *
 */

public class Dictionary implements DictionaryInterface  {

    // blackbox, private Node class:
    private class Node {
        String key;
        String value;
        Node next;

        Node(String key, String value) {
            this.key = key;
            this.value = value;
            next = null;
        }
    }
    
    // Dictionary fields:
    // head: is a reference to first Node in List
    // tail: is a reference to last Node in list
    // numItems: # of items in the dictionary
    private Node head;
    private Node tail;  
    private int numItems;
    

    // Dictionary Constructor:
    public Dictionary () {
        head = null;
        tail = null;
        numItems = 0;

    }
   
    // findKey()-------------------------------------------------------------------------
    // pre: Head != null
    // post: returns a reference to the Node with this key inside the dictionary
    private Node findKey(String key) {
        Node N = head;
        while( N != null) {
          if(key.compareTo(N.key) == 0) {
               return N;
           }
           else {
               N = N.next;
          }
        }
        return null;
    }

    // ADT Operations ===================================================================

    // isEmpty()-------------------------------------------------------------------------
    // pre: none
    // post: returns true if Dictionary contains no pairs (is empty), false otherwise
    public boolean isEmpty() {
        return(numItems == 0);
    }    
    

    // size()---------------------------------------------------------------------------
    // pre: none
    // post: returns number of (key, value) pairs in the Dictionary
    public int size() {
        return numItems;
    }
    
    // lookup()-------------------------------------------------------------------------
    // pre: none
    // post: returns pair if such pair is contained within the dictionary. 
    //       and null if no such key exists.
     public String lookup(String key) {  
        Node N = findKey(key);
        if( N == null) {
            return null;
        }
        return N.value;
     }

    // insert() 
    // pre: key must not already exist in the Dictionary: lookup(key) == null
    // post: inserts new (key,value) pair into the Dictionary
    public void insert(String key, String value) throws DuplicateKeyException {
        if (lookup(key) != null ) {
            throw new DuplicateKeyException("insert(): cannot insert duplicate keys" +key); 
        }
        if(head == null) {
            Node N = new Node(key,value);
            head = N;
            tail = N;
        }
        else if(head == tail) {
            Node N = head;
            N.next = new Node(key,value);
            tail = N.next;
        }
        else {
           Node N = tail;
           N.next = new Node(key,value);
           tail = N.next;
        }
        numItems++;
    }

    
    // delete()-------------------------------------------------------------------------
    // pre: checks to see if the key is there: lookup(key) != null
    // post: deletes pair with the give key value
    public void delete(String key) throws KeyNotFoundException {
        if(lookup(key) == null) {
           throw new KeyNotFoundException("delete(): cannot delete a non-existent key" + key);
        }
        else if(findKey(key) == head) {
            if(head == tail) {
                head = null;
                tail = null;
            }
            else {
                Node N = head;
                head = N.next;
                N = null;
            }
        }
        else if(findKey(key) == tail) {
            Node S = tail;
            Node P = head;
            while(P.next != tail) {
                P = P.next;
            }
            P.next = null;
            tail = P;
        }
        else {
           Node S = findKey(key);
           Node P = head;
           while(P.next != S) {
               P = P.next;
           }
           P.next = S.next;
           S = null;
        }
        numItems--;       
    }    

     
    // makeEmpty()-------------------------------------------------------------------------
    // pre: none
    // post: Resets the Dictionary to the empty state
    public void makeEmpty() {
        head = null;
        tail = null;
        numItems = 0;
    } 

    // toString()--------------------------------------------------------------------------
    // pre: none
    // post: prints the current state to stdout, Overides Object's toString()
    public String toString() {
        String s = "";
        for (Node N = head; N != null; N = N.next) {
            s += N.key + " " + N.value + "\n";
        }
        return s;
    }
}

